Index: simplesamlphp-1.14.11/lib/SimpleSAML/Utils/HTTP.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- simplesamlphp-1.14.11/lib/SimpleSAML/Utils/HTTP.php	(date 1492191413000)
+++ simplesamlphp-1.14.11/lib/SimpleSAML/Utils/HTTP.php	(revision )
@@ -98,6 +98,8 @@
      */
     private static function getServerPort()
     {
+        // Acquia specific change.
+        return '';
         $port = (isset($_SERVER['SERVER_PORT'])) ? $_SERVER['SERVER_PORT'] : '80';
         if (self::getServerHTTPS()) {
             if ($port !== '443') {
