diff --git a/simple_menu_icons.module b/simple_menu_icons.module
index a978b31..1e87b27 100755
--- a/simple_menu_icons.module
+++ b/simple_menu_icons.module
@@ -140,9 +140,9 @@ function simple_menu_icons_css_generate() {
 }
 
 /**
- * Implements hook_cache_flush().
+ * Implements hook_rebuild().
  */
-function simple_menu_icons_cache_flush() {
+function simple_menu_icons_rebuild() {
   simple_menu_icons_css_generate();
 }
 
