diff --git a/page_access.module b/page_access.module
index 6dc4cff..a85ce4a 100644
--- a/page_access.module
+++ b/page_access.module
@@ -329,7 +329,7 @@ function page_access_node_access(\Drupal\node\NodeInterface $node, $op, \Drupal\
         return AccessResult::allowed();
       }
       else {
-        return AccessResult::forbidden();
+        return AccessResult::neutral();
       }
 
     default:
