diff --git a/page_access.module b/page_access.module
index 6dc4cff..d78abc8 100644
--- a/page_access.module
+++ b/page_access.module
@@ -30,7 +30,7 @@ function page_access_form_node_form_alter(&$form, FormStateInterface &$form_stat
     '#type' => 'details',
     '#title' => t('Page access settings'),
     '#access' => \Drupal::currentUser()->hasPermission('administer menu'),
-    '#open' => (bool) $defaults['id'],
+    '#open' => (isset($defaults['id']) ? (bool) $defaults['id'] : ''),
     '#group' => 'advanced',
     '#attached' => [
       'library' => ['page_access/drupal.page_access.adminforms'],
@@ -40,12 +40,12 @@ function page_access_form_node_form_alter(&$form, FormStateInterface &$form_stat
   ];
   $form['page_access_entity_id'] = [
     '#type' => 'hidden',
-    '#value' => $defaults['id'],
+    '#value' => (isset($defaults['id']) ? $defaults['id'] : ''),
   ];
   $form['page_access']['enabled'] = [
     '#type' => 'checkbox',
     '#title' => t('Enable page access control'),
-    '#default_value' => (int) (bool) $defaults['id'],
+    '#default_value' => (isset($defaults['id']) ? (int) (bool) $defaults['id'] : ''),
   ];
   $form['page_access']['roles'] = [
     '#type' => 'checkboxes',
@@ -57,7 +57,7 @@ function page_access_form_node_form_alter(&$form, FormStateInterface &$form_stat
         'input[name="page_access[enabled]"]' => ['checked' => FALSE],
       ],
     ],
-    '#default_value' => $defaults['page_access']['roles'],
+    '#default_value' => (isset($defaults['page_access']['roles'])) ? $defaults['page_access']['roles'] : array(),
   ];
   $form['page_access']['users'] = [
     '#type' => 'item',
@@ -95,7 +95,11 @@ function page_access_form_node_form_alter(&$form, FormStateInterface &$form_stat
       unset($form['page_access']['list'][$i]);
       continue;
     }
-    $user_reference = _get_user_reference_default_value($defaults['page_access']['users'][$i]['user_autocomplete']);
+    if(isset($defaults['page_access']['users'])) {
+      $user_reference = _get_user_reference_default_value($defaults['page_access']['users'][$i]['user_autocomplete']);
+    } else {
+      $user_reference = '';
+    }
     $form['page_access']['list'][$i]['user_autocomplete'] = [
       '#type' => 'entity_autocomplete',
       '#target_type' => 'user',
@@ -107,12 +111,17 @@ function page_access_form_node_form_alter(&$form, FormStateInterface &$form_stat
       '#prefix' => '<div class="container-inline" id="page-access-' . $i . '">',
       '#default_value' => $user_reference,
     ];
+    if(isset($defaults['page_access']['users'])) {
+      $edit_permission = (int) (bool) $defaults['page_access']['users'][$i]['edit_permission'];
+    } else {
+      $edit_permission = '';
+    }
     $form['page_access']['list'][$i]['edit_permission'] = [
       '#type' => 'checkbox',
       '#default_value' => 0,
       '#title' => 'Edit',
       '#attributes' => ['class' => ['page-access-edit-access']],
-      '#default_value' => (int) (bool) $defaults['page_access']['users'][$i]['edit_permission'],
+      '#default_value' => $edit_permission,
     ];
     if ($i) {
       $form['page_access']['list'][$i]['remove'] = [
diff --git a/src/Entity/PageAccess.php b/src/Entity/PageAccess.php
index 4e2f3cc..3ba75ef 100644
--- a/src/Entity/PageAccess.php
+++ b/src/Entity/PageAccess.php
@@ -80,9 +80,10 @@ class PageAccess extends ContentEntityBase implements PageAccessInterface {
    *
    * @return \Drupal\Core\Entity\EntityInterface[]
    */
-  public function loadByNid($nid) {
+  public static function loadByNid($nid) {
     $page_access_storage = \Drupal::entityManager()->getStorage('page_access');
-    return array_shift($page_access_storage->loadByProperties(['nid' => $nid]));
+    $page_access_loaded_entities = $page_access_storage->loadByProperties(['nid' => $nid]);
+    return array_shift($page_access_loaded_entities );
   }

   /**
