diff --git a/core/lib/Drupal/Core/Path/PathValidator.php b/core/lib/Drupal/Core/Path/PathValidator.php
index f7efdb1..b50b90c 100644
--- a/core/lib/Drupal/Core/Path/PathValidator.php
+++ b/core/lib/Drupal/Core/Path/PathValidator.php
@@ -14,6 +14,7 @@
 use Symfony\Component\Routing\Exception\MethodNotAllowedException;
 use Symfony\Component\Routing\Exception\ResourceNotFoundException;
 use Symfony\Component\Routing\Matcher\UrlMatcherInterface;
+use Symfony\Component\Routing\RequestContext;
 
 /**
  * Provides a default path validator and access checker.
@@ -145,11 +146,15 @@ protected function getUrl($path, $access_check) {
    *   An array of request attributes of FALSE if an exception was thrown.
    */
   protected function getPathAttributes($path, Request $request, $access_check) {
+    $requestContext = new RequestContext();
+    $requestContext->fromRequest($request);
     if (!$access_check || $this->account->hasPermission('link to any page')) {
       $router = $this->accessUnawareRouter;
+      $router->setContext($requestContext);
     }
     else {
       $router = $this->accessAwareRouter;
+      $router->setContext($requestContext);
     }
 
     $path = $this->pathProcessor->processInbound('/' . $path, $request);
