diff --git a/docroot/core/modules/content_moderation/src/ModerationInformation.php b/docroot/core/modules/content_moderation/src/ModerationInformation.php
index ed33902..0a37ee9 100644
--- a/docroot/core/modules/content_moderation/src/ModerationInformation.php
+++ b/docroot/core/modules/content_moderation/src/ModerationInformation.php
@@ -81,6 +81,7 @@ public function getLatestRevisionId($entity_type_id, $entity_id) {
         ->condition($this->entityTypeManager->getDefinition($entity_type_id)->getKey('id'), $entity_id)
         ->sort($this->entityTypeManager->getDefinition($entity_type_id)->getKey('revision'), 'DESC')
         ->range(0, 1)
+        ->accessCheck(FALSE)
         ->execute();
       if ($revision_ids) {
         return array_keys($revision_ids)[0];
