Index: docroot/core/modules/views/src/Controller/ViewAjaxController.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- docroot/core/modules/views/src/Controller/ViewAjaxController.php	(date 1481383904000)
+++ docroot/core/modules/views/src/Controller/ViewAjaxController.php	(revision )
@@ -192,6 +192,8 @@
         }
         $response->addCommand(new ReplaceCommand(".js-view-dom-id-$dom_id", $preview));
 
+        \Drupal::moduleHandler()->alter('views_ajax_response', $response, $view);
+
         return $response;
       }
       else {
