Index: docroot/modules/contrib/xmlsitemap/src/XmlSitemapIndexWriter.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- docroot/modules/contrib/xmlsitemap/src/XmlSitemapIndexWriter.php	(date 1480678768000)
+++ docroot/modules/contrib/xmlsitemap/src/XmlSitemapIndexWriter.php	(revision )
@@ -52,7 +52,7 @@
     for ($i = 1; $i <= $this->sitemap->chunks; $i++) {
       $url_options['query']['page'] = $i;
       $element = array(
-        'loc' => Url::fromRoute('xmlsitemap.sitemap_xml', [], $url_options),
+        'loc' => Url::fromRoute('xmlsitemap.sitemap_xml', [], $url_options)->toString(),
         // @todo Use the actual lastmod value of the chunk file.
         'lastmod' => gmdate($lastmod_format, REQUEST_TIME),
       );
